README.md
<!----- Conversion time: 0.718 seconds.
Using this Markdown file:

1. Cut and paste this output into your source file.
2. See the notes and action items below regarding this conversion run.
3. Check the rendered output (headings, lists, code blocks, tables) for proper
   formatting and use a linkchecker before you publish this page.

Conversion notes:

* Docs to Markdown version 1.0β17
* Wed Sep 18 2019 01:52:00 GMT-0700 (PDT)
* Source doc: https://docs.google.com/open?id=1SEODmwLcgVdQijJMZ6Xc3YQ0lqnkc72w-gccG4AkpqU
----->

## Практикум

### Цель работы

Познакомиться с приемами работы с многопоточностью на примере создания сокетного TCP-сервера, способного работать с несколькими клиентами одновременно

### Контрольные вопросы

1. Почему однопоточное приложение не может решить задачу одновременного подключения? Потому что программа будет занята либо обслуживанием уже подключившегося пользователя, либо будет принимать соединение от нового. Иначе говоря, в один момент времени программа выполняет только одно "действие".

2. Чем поток отличается от процесса?Процессы и потоки связаны друг с другом, но при этом имеют существенные различия. Процесс — экземпляр программы во время выполнения, независимый объект, которому выделены системные ресурсы (например, процессорное время и память). Каждый процесс выполняется в отдельном адресном пространстве: один процесс не может получить доступ к переменным и структурам данных другого. Если процесс хочет получить доступ к чужим ресурсам, необходимо использовать межпроцессное взаимодействие. Это могут быть конвейеры, файлы, каналы связи между компьютерами и многое другое. Поток использует то же самое пространства стека, что и процесс, а множество потоков совместно используют данные своих состояний. Как правило, каждый поток может работать (читать и писать) с одной и той же областью памяти, в отличие от процессов, которые не могут просто так получить доступ к памяти другого процесса. У каждого потока есть собственные регистры и собственный стек, но другие потоки могут их использовать. Поток — определенный способ выполнения процесса. Когда один поток изменяет ресурс процесса, это изменение сразу же становится видно другим потокам этого процесса.

3. Как создать новый поток? 
4. Как выделить участок кода так, чтобы он выполнялся в другом потоке?

![image](https://user-images.githubusercontent.com/70951761/139743420-1e96b126-380e-420f-bc33-cca5bea296b4.png)
![image](https://user-images.githubusercontent.com/70951761/139743537-e102fe49-1b10-4a05-8c8f-7c00ddd3c234.png)

5. В чем проблема потокобезопасности?С точки зрения безопасности, почему мы заботимся о безопасности потоков? Потому что у безопасности памяти и потоков одна и та же основная проблема: недопустимое использование ресурсов. Здесь атаки приводят к таким же последствиям, как атаки на память, включая повышение привилегий, выполнение произвольного кода (ACE) и обход проверок безопасности.

6. Какие методы обеспечения потокобезопасности существуют?Класс является потокобезопасным, если он ведет себя правильно при доступе из нескольких потоков, независимо от планирования или чередования выполнения этих потоков средой выполнения, и без дополнительной синхронизации или другой координации со стороны вызывающего кода. Простейший способ синхронизации с общими переменными или другими ресурсами, это положиться на атомарные операции в интерпретаторе. Атомарная операция – это операция, которая осуществляет только один этап выполнения задачи, без каких либо шансов того, что другой поток получит контроль.Блокировки – это фундаментальный механизм синхронизации, который предоставлен модулем threading Python. Замок может удерживаться одним потоком в любое время, или без потока вообще. Если поток попытается удержать один замок, который уже удерживается другим потоком, выполнение первого потока будет остановлена, пока не будет снята блокировка. Замки обычно используются для синхронизации доступа к общим ресурсам. Для каждого такого источника создается объект Lock.Класс RLock – это версия замка, который выполняет функцию блокировки только в том случае, если замок удерживает другой поток. В то время как обычные замки блокируют тогда, когда тот же поток пытается получить к одному и тому же замку дважды, реентерабельный замок блокирует только в том случае, если другой поток уже держит замок.Семафор — это более продвинутый механизм блокировки. В семафоре есть внутренний счетчик, вместо флага блокировки, и блокирует он только в том случае, если большее количество потоков (чем указано) попытается удержать семафор. В зависимости от того, как инициализирован семафор, это позволяет нескольким потокам получить доступ к одной и той же части кода одновременно.Объект event – это простой объект синхронизации. Он представляет собой внутренний флаг, так что все потоки могут ожидать, пока флаг будет установлен, задавать, или убирать его.Объект condition это более продвинутая версия объекта event. Он представляет собой своего рода измененное состояние в приложении, и поток может дожидаться заданных условий, или сигнал о том, что условие было задано.

### Дополнительные задания

1. Модифицировать простой эхо-сервер таким образом, чтобы при подключении клиента создавался новый поток, в котором происходило взаимодействие с ним.

![image](https://user-images.githubusercontent.com/70951761/139740480-951ee723-f981-44bf-adcf-393fcfe55ddd.png)

2. Реализовать простой чат сервер на базе сервера аутентификации. Сервер должен обеспечивать подключение многих пользователей одновременно, отслеживание имен пользователей, поддерживать историю сообщений и пересылку сообщений от каждого пользователя всем остальным. 

![image](https://user-images.githubusercontent.com/70951761/139741230-384fc346-ea73-4d04-9347-51243f7b0f96.png)

3. Реализовать сервер с управляющим потоком. При создании сервера прослушивание портов происходит в отдельном потоке, а главный поток программы в это время способен принимать команды от пользователя. Необходимо реализовать следующие команды:
    1. Отключение сервера (завершение программы);
    2. Пауза (остановка прослушивание порта);
    3. Показ логов;
    4. Очистка логов;
    5. Очистка файла идентификации.
 
![image](https://user-images.githubusercontent.com/70951761/139742471-7aee23b3-e1e7-4b28-bf6d-4b56c8018156.png)
![image](https://user-images.githubusercontent.com/70951761/139742507-4987f562-8fd5-40a4-ad15-d1b903b8fe33.png)

<!-- Docs to Markdown version 1.0β17 -->


history.txt
bob:hi!
jack:hello!
bob:how are u?
jack:i am fine, and u?
bob:me too)


client_pr.py
import socket
import getpass
import re
import threading
import os

DEF_HOST = '127.0.0.1'  # The server's hostname or IP address
DEF_PORT = 65432        # The port used by the server
max = 100000
tokenfile = 'token.txt'


def user_conf():
    ip = re.search(r'^\d{1,3}[.]\d{1,3}[.]\d{1,3}[.]\d{1,3}$', getpass.getpass(
        prompt='enter ip you want to connect: '))
    ip = ip.group() if ip else DEF_HOST
    while True:
        user_port = getpass.getpass(
            prompt="Enter port you want to use: ", stream=None)
        try:
            user_port = int(
                user_port) if user_port != '' else DEF_PORT
            user_port = user_port if 1 < user_port < 65537 else DEF_PORT
            break
        except:
            continue
    return ip, user_port


def receive(conn):
    while True:
        try:
            message = conn.recv(1024).decode()
            print(message)
        except:
            break


def token():
    if tokenfile in os.listdir(os.getcwd()):
        with open(tokenfile, 'r', encoding='utf-8') as ss:
            t = ss.read()
            return t if t else 'no token'
    else:
        a = open(tokenfile, 'w')
        a.close()
        return 'no token'


def write_token(line):
    with open(tokenfile, 'w', encoding='utf-8') as file:
        file.write(line)


with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
    ip, port = user_conf()
    try:
        s.connect((ip, port))  # connect to server
        print('connection with server was successfull')
    except:
        print('Connection can not be complete')
        exit()
    while True:
        sign = s.recv(1024).decode()
        if 'name' in sign:
            s.send(input('Enter your name: ').encode())
        elif 'password' in sign:
            name = sign.split(',')[1]
            s.send(input(f'Enter your password, {name}: ').encode())
        elif 'token' in sign:
            tk = token()
            s.send(tk.encode())
        else:
            name = sign.split(',')[1]
            t = sign.split(',')[2] if len(sign.split(',')) > 2 else False
            if t:
                write_token(t)
            break
    print('----------enter for exit----------')
    data = s.recv(max).decode()
    if data == 'no messages':
        pass
    else:
        data = ''.join([i.lstrip()
                        for i in re.split(rf'{name}[:]', data)])[:-1]
        print(data)
    a = threading.Thread(target=receive, args=(s,))
    a.start()
    while True:
        text = str(input())
        if not text:
            break
        s.send(text.encode())


server_pr.py
import os
import socket
import getpass
import csv
import threading
import datetime


class Server():
    listen_flag = True
    key = 'sglypa)'
    threadlist = []
    history = 'history.txt'
    logfile = 'log.txt'
    usernames = []
    log_text = {1: 'server is starting', 2: 'port is listened',
                3: 'connection was successfull', 4: 'server got data',
                5: 'client was disconnected', 6: 'server was turned off',
                7: 'server changed port', 8: 'new client', 9: 'message',
                10: 'pause', 11: 'show logs', 12: 'clear fileident'}
    Def_HOST = '127.0.0.1'  # Standard loopback interface address (localhost)
    Def_PORT = 65432    # Port to listen on (non-privileged ports are > 1023)
    commands = ['listen', 'off', 'pause',
                'show logs', 'clear logs', 'clear fileident', '?']

    def __init__(self, open_port, host):
        self.open_port = open_port
        self.host = host

    def change_port(self, port):
        self.open_port = port

    @staticmethod
    def log(code):
        with open(Server.logfile, 'a', encoding='utf-8') as file:
            if type(code) == int:
                file.write(Server.log_text[code] + '\t' +
                           str(datetime.datetime.now())+'\n')
            else:
                file.write(code+'\t'+str(datetime.datetime.now())+'\n')

    @staticmethod
    def vernam_enc_dec(k, m):
        k = k*(len(m)//len(k)) + k[-(len(m) % len(k)):]
        return ''.join(map(chr, [i ^ x for i, x in zip(map(ord, m), map(ord, k))]))

    @staticmethod
    def start_program():
        User.create_file()
        while True:
            user_port = getpass.getpass(
                prompt="Enter port you want to use: ", stream=None)
            try:
                user_port = int(
                    user_port) if user_port != '' else Server.Def_PORT
                user_port = user_port if 1 < user_port < 65537 else Server.Def_PORT
                break
            except:
                continue
        a = Server(user_port, Server.Def_HOST)
        a.run()

    @staticmethod
    def listening(sock):
        while True:
            if Server.listen_flag:
                try:
                    conn, addr = sock.accept()
                except:
                    break
                thread = ClientThread(len(Server.threadlist), conn, addr)
                Server.threadlist.append(thread)
                thread.start()
            else:
                continue

    def run(self):
        Server.log(1)
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            while True:
                try:
                    s.bind((self.host, self.open_port))
                    break
                except:
                    self.change_port(self.open_port+1)
                    Server.log(7)
            s.listen(5)
            thread = threading.Thread(
                target=Server.listening, args=(s,))
            Server.threadlist.append(thread)
            thread.start()
            Server.log(2)
            print(f'listen {self.open_port}')
            while True:
                command = input('enter the command (? for commands list): ')
                if command == 'off':
                    Server.log(6)
                    for n, c in Server.usernames:
                        c.close()
                    raise SystemExit
                elif command == '?':
                    print(', '.join(Server.commands))
                elif command == 'listen':
                    Server.listen_flag = True
                    Server.log(2)
                    print(f'listen {self.open_port}')
                elif command == 'pause':
                    Server.listen_flag = False
                    Server.log(10)
                    print(f'pause {self.open_port}')
                elif command == 'clear fileident':
                    Server.log(12)
                    with open(User.for_users, 'w', encoding='utf-8') as ss:
                        pass
                elif command == 'clear logs':
                    print('clear logs...')
                    with open(Server.logfile, 'w', encoding='utf-8') as ss:
                        pass
                elif command == 'show logs':
                    Server.log(11)
                    with open(Server.logfile, 'r', encoding='utf-8') as ss:
                        text = ss.read()
                        print(text)


class User():
    for_users = 'saved_users.csv'
    userlist = []

    @staticmethod
    def create_file():
        if User.for_users in os.listdir(os.getcwd()):
            with open(User.for_users, encoding='utf-8') as s:
                reader = csv.reader(s, delimiter=';')
                User.userlist = [row for row in reader]
        else:
            a = open(User.for_users, 'w', encoding='utf-8')
            a.close()

    @staticmethod
    def sync():
        with open(User.for_users, 'w', encoding='utf-8') as s:
            writer = csv.writer(s, delimiter=';')
            writer.writerows(User.userlist)

    @staticmethod
    def token():
        from random import randint
        return ''.join([chr(randint(1, 2000)) for i in range(40)])


class ClientThread(threading.Thread):
    def __init__(self, name, connector, addr):
        threading.Thread.__init__(self)
        self.name = name
        self.connector = connector
        self.clientaddr = addr

    def run(self):
        name = ClientThread.ident(self.connector)
        Server.usernames.append((name, self.connector))
        Server.log('identification')
        self.message_hist()
        Server.log('message history was sent')
        while True:
            text = ClientThread.recv_(self.connector)
            if not text:
                break
            ClientThread.send_(text, name)

    @staticmethod
    def new_user(sock):
        sock.send('name'.encode())
        name = sock.recv(1024).decode()
        sock.send(f'password,{name}'.encode())
        answer = sock.recv(1024).decode()
        password = Server.vernam_enc_dec(
            Server.key, answer)
        token = User.token()
        sock.send(f'Hello,{name},{token}'.encode())
        User.userlist.append([name, password, token])
        User.sync()
        return name

    @ staticmethod
    def ident(sock):
        if len(User.userlist) == 0:
            return ClientThread.new_user(sock)
        else:
            sock.send('token'.encode())
            token = sock.recv(1024).decode()
            for row in User.userlist:
                if row[2] == token:
                    sock.send(f'Hello,{row[0]}'.encode())
                    return row[0]
            else:
                sock.send('name'.encode())
                name = sock.recv(1024).decode()
                for i, row in enumerate(User.userlist):
                    if row[0] == name:
                        while True:
                            sock.send(f'password,{name}'.encode())
                            passwd = sock.recv(1024).decode()
                            data = Server.vernam_enc_dec(
                                Server.key, passwd)
                            if data == row[1]:
                                token = User.token()
                                sock.send(f'Hello,{row[0]},{token}'.encode())
                                User.userlist[i].pop()
                                User.userlist[i].append(token)
                                User.sync()
                                return row[0]
                else:
                    sock.send(f'password,{name}'.encode())
                    answer = sock.recv(1024).decode()
                    password = Server.vernam_enc_dec(
                        Server.key, answer)
                    token = User.token()
                    sock.send(f'Hello,{name},{token}'.encode())
                    User.userlist.append([name, password, token])
                    User.sync()
                    return name

    @staticmethod
    def send_(message, name):
        for name_, conn in Server.usernames:
            if name_ != name:
                conn.send(f'{name}:{message}'.encode())
        with open(Server.history, 'a', encoding='utf-8') as a:
            a.write(f'{name}:{message}\n')

    @staticmethod
    def recv_(conn):
        try:
            text = conn.recv(1024)
            return text.decode()
        except:
            pass

    def message_hist(self):
        if Server.history in os.listdir(os.getcwd()):
            with open(Server.history, 'r', encoding='utf-8') as ss:
                a = ss.read()
                if len(a) == 0:
                    self.connector.send('no messages'.encode())
                else:
                    self.connector.send(f'{a}'.encode())
        else:
            a = open(Server.history, 'w', encoding='utf-8')
            self.connector.send('no messages'.encode())
            a.close()


if __name__ == '__main__':
    Server.start_program()
